renovate/src/Renovate/BinaryFormat/ELF.hs

```diff
@@ -465,19 +494,126 @@ doRewrite cfg hdlAlloc loadedBinary symmap strat = do
   -- they're uninteresting.
   modifyCurrentELF (\e -> ((),) <$> E.traverseElfSegments nullifyPhdr e)
   modifyCurrentELF (appendSegment newSegment)
+
+  -- Linux calculates the program break based on the virtual address and size of
+  -- the LOAD segment with the highest virtual address, whereas QEMU calculates
+  -- it based on the virtual address and size of the *writable* LOAD segment with
+  -- the highest virtual address. Since our PHDR segment might have a high
+  -- virtual address, we have to add a writable load segment after it, so our
+  -- programs don't segfault inside QEMU.
+  --
+  -- https://bugs.launchpad.net/qemu/+bug/1886097
+
+  phdrActualSize <-
+    withCurrentELF $ \elf ->
+      E.phdrMemSize <$>
+        case filter ((== E.PT_PHDR) . E.phdrSegmentType) (E.allPhdrs (E.elfLayout elf)) of
+          [seg] -> pure seg
+          phdrs' ->
+            fail $ unwords $
+              [ "Internal error:"
+              , "Wrong number of PT_PHDR segments:"
+              , show phdrs'
+              ]
+
+  nextIdx <- withCurrentELF (pure . (+1) . nextSegmentIndex)
+  modifyCurrentELF $ appendSegment $
+    let alignedAddr =
+          alignValue
+            (E.elfSegmentVirtAddr newSegment + phdrActualSize)
```

Comment:
The comment above this section says:
```
  -- the highest virtual address. Since our PHDR segment might have a high
  -- virtual address, we have to add a writable load segment after it, so our
  -- programs don't segfault inside QEMU.
```
But I will also add a line comment here specifically.

Reply:
https://github.com/GaloisInc/renovate/pull/55/commits/7d80396db4fe23e9ae4a6592450f7963dcb253e3